insights/core/spec_factory.py

```diff
@@ -736,6 +736,62 @@ def __call__(self, broker):
                 keep_rc=self.keep_rc, ds=self, timeout=self.timeout, inherit_env=self.inherit_env)
 
 
+class command_with_args(object):
+    """
+    Execute a command that has dynamic arguments
+
+    Args:
+        cmd (list of lists): the command to execute. Breaking apart a command
+            string that might require arguments.
+        provider (str or list): argument string or a list of arguments.
+        context (ExecutionContext): the context under which the datasource
+            should run.
+        split (bool): whether the output of the command should be split into a
+            list of lines
+        keep_rc (bool): whether to return the error code returned by the
+            process executing the command. If False, any return code other than
+            zero with raise a CalledProcessError. If True, the return code and
+            output are always returned.
+        timeout (int): Number of seconds to wait for the command to complete.
+            If the timeout is reached before the command returns, a
+            CalledProcessError is raised. If None, timeout is infinite.
+        inherit_env (list): The list of environment variables to inherit from the
+            calling process when the command is invoked.
+
+    Returns:
+        function: A datasource that returns the output of a command that takes
+            specified arguments passed by the provider.
+    """
+
+    def __init__(self, cmd, provider, context=HostContext, deps=None, split=True, keep_rc=False, timeout=None, inherit_env=None, **kwargs):
+        deps = deps if deps is not None else []
+        self.cmd = cmd
+        self.provider = provider
+        self.context = context
+        self.split = split
+        self.raw = not split
+        self.keep_rc = keep_rc
+        self.timeout = timeout
+        self.inherit_env = inherit_env if inherit_env is not None else []
+        self.__name__ = self.__class__.__name__
+        datasource(self.provider, self.context, *deps, raw=self.raw, **kwargs)(self)
+
+    def __call__(self, broker):
+        source = broker[self.provider]
+        ctx = broker[self.context]
+        if isinstance(source, ContentProvider):
+            source = source.content
+        if not isinstance(source, (list, set, tuple)):
+            source = [source]
+        try:
+            self.cmd = self.cmd % ' '.join(source)
```

Comment:
Instead of `' '.join(source)`, let's do something like `self.cmd % tuple(source)`. That way substitution parameters can appear anywhere in `self.cmd`.

We should also only accept `list` and `tuple` instances since `set` would make the string substitution ambiguous.

Reply:
@csams - Please correct me if I misunderstood you. 
By replacing the above  `' '.join(source)`  with `tuple(source)`, it requires the source datasource to pass a list(or tuple) with the specified number of strings exactly match with the request of the `self.cmd`. And that requires each item in the passed list(or tuple) should be a single string (but not a list). Is it right? e.g.

```python
@datasource(...)
def ds_1_src(lst):
     return ','.join(lst)
@datasource(...)
def ds_2_src(lst):
     return [','.join(lst), "test"]
@datasource(...)
def ds_3_src(lst, lst2):
     return [','.join(lst), "test", ' '.join(lst2)]

ds_1_args = command_with_args("/bin/xxx -a %s", ds_1_src)
ds_2_args = command_with_args("/bin/xxx -a %s -b %s", ds_2_src)
ds_3_args = command_with_args("/bin/xxx -a %s -b %s -c %s", ds_3_src)
```